### PR TITLE
Fix similar to only return SentinelArray when new type includes sentinel

### PR DESCRIPTION
**Pull request by quinnj**
type

### CODE REVIEW

**Comment by bkamins** [comment]
Does this solve this problem:
```
julia> x
2-element SentinelArrays.SentinelArray{Float64,1,Float64,Missing,Array{Float64,1}}:
  missing
 0.6341068459430321

julia> eltype(x)
Union{Missing, Float64}

julia> similar(x, Float64)
2-element SentinelArrays.SentinelArray{Float64,1,Float64,Missing,Array{Float64,1}}:
 missing
 missing

julia> eltype(similar(x, Float64))
Union{Missing, Float64}
```
(this should be an error)

while this should work and make eltype `Float64`:
```
julia> similar(x[2:2], Float64)
1-element SentinelArrays.SentinelArray{Float64,1,Float64,Missing,Array{Float64,1}}:
 missing

julia> eltype(similar(x[2:2], Float64))
Union{Missing, Float64}
```

File: test/runtests.jl [hunk 1]
```diff
@@ -35,14 +35,14 @@ y = similar(x, 20)
...
 @test y[1] === missing
-y = similar(x, Float64, 10)
+y = similar(x, Union{Missing, Float64}, 10)
 @test eltype(parent(y)) === Float64
 @test y[1] === missing
 @test length(y) == 10
-y = similar(x, Float64, (10, 10))
+y = similar(x, Union{Missing, Float64}, (10, 10))
 @test eltype(parent(y)) === Float64
 @test y[1] === missing
```
Comment (nalimilan) [inline test/runtests.jl:47]: Doesn't this PR need a new test? None of the existing ones appear to cover a non-missing eltype.

Comment (bkamins) [reply test/runtests.jl:47]: These tests are exactly ones that I have added in the comment above.

Comment (quinnj) [reply test/runtests.jl:47]: Added a test in 0a713b158f3f81fdeef32c234943b64004d8c9ce
